### PR TITLE
AP_BattMonitor: Add smart battery cell voltage minimum and diff monitoring

### PR DESCRIPTION
**Pull request by Pedals2Paddles**
This adds checking of individual cell voltages to the low/critical failsafes.  The low and critical battery failsafes will trigger if the lowest cell voltage drops below a threshold.  And the low battery failsafe will also trigger if the difference between the lowest and highest cell voltage increases beyond a threshold. I extended the methods used for the existing checks into the cell voltage and difference.

And.... it works!

_Updated after discussing with MDB_
- Adds parameters `_CELL_V_LOW`, `_CELL_V_CRT`, & `_CELL_V_DIF`.  The default for all 3 is 0, which is disabled.  These are the threshold voltages for low, critical, and high-to-low difference.  Low cell would typically be 3.45v, critical cell 3.30v, and diff I think maybe 0.15 or something.

- Adds checking the lowest cell voltage to the low and critical battery failsafes.  Adds checking the difference between the highest and lowest cell to the **low** battery failsafe.  It didn't make sense to add a whole other type of failsafe for this.  If the diff is too great, one of the cells is obviously too low, so it is an appropriate place for it.  

To-Do: Cell data is too much for the existing send_text that goes out when the failsafe trips.  Needs something else.

### CODE REVIEW

**Comment by Pedals2Paddles** [comment]
Travis is failing only on the **CI_BUILD_TARGET="px4-v2 revo-bootloader iofirmware"** and the error is  `region `flash' overflowed by 100 bytes`.

File: libraries/AP_BattMonitor/AP_BattMonitor.h [hunk 1]
```diff
@@ -78,6 +78,8 @@ class AP_BattMonitor
         float       resistance;                // resistance, in Ohms, calculated by comparing resting voltage vs in flight voltage
         BatteryFailsafe failsafe;              // stage failsafe the battery is in
         bool        healthy;                   // battery monitor is communicating correctly
+        float       cell_min;                  // lowest smart battery cell voltage
```
Comment (WickedShell) [inline libraries/AP_BattMonitor/AP_BattMonitor.h]: I'd keep this out of state, it's increasing the general memory footprint for all backends, and assuming a 10Hz SMBus smart battery (which is what we currently do) this isn't actually helping you save processing time. It's also forcing a more complicated getter relationship then you really need.

File: libraries/AP_BattMonitor/AP_BattMonitor.cpp [hunk 2]
```diff
@@ -369,8 +387,14 @@ void AP_BattMonitor::check_failsafes(void)
                     break;
             }
 
-            gcs().send_text(MAV_SEVERITY_WARNING, "Battery %d is %s %.2fV used %.0f mAh", i + 1, type_str,
-                            (double)voltage(i), (double)consumed_mah(i));
+            if (has_cell_voltages(i)) {
+                gcs().send_text(MAV_SEVERITY_WARNING, "Batt %d %s; %.2fV; Used %.0fmAh; Cell %.2fV Diff %.2fV; ", i + 1, type_str,
```
Comment (WickedShell) [inline libraries/AP_BattMonitor/AP_BattMonitor.cpp]: This string is much to long, the entire formatted string needs to fit in 50 characters. That should be computed using the longest failsafe type (`critical`) and a battery voltage >10V, and used capacity >1000 mAh (>10k is better).

File: libraries/AP_BattMonitor/AP_BattMonitor_SMBus_Solo.cpp [hunk 3]
```diff
@@ -43,10 +43,16 @@ void AP_BattMonitor_SMBus_Solo::timer()
     // read cell voltages
     if (read_block(BATTMONITOR_SMBUS_SOLO_CELL_VOLTAGE, buff, 8, false)) {
         float pack_voltage_mv = 0.0f;
+        float cell_diff_mv = 0.0f;
```
Comment (WickedShell) [inline libraries/AP_BattMonitor/AP_BattMonitor_SMBus_Solo.cpp]: Don't do this here, this logic is identical for all the SMBus batteries. This should either be done as a single loop in the general SMBus backend, or just computed on demand in the failsafe check (this would be my preference as we could remove the state then).

File: libraries/AP_BattMonitor/AP_BattMonitor_Backend.cpp [hunk 4]
```diff
@@ -214,4 +216,28 @@ void AP_BattMonitor_Backend::check_failsafe_types(bool &low_voltage, bool &low_c
...
+    if (has_cell_voltages()) {
+
+        // critical cell voltage
+        if ((_params._crt_cell_voltage > 0) && (_state.cell_min < _params._crt_cell_voltage)) {
+            critical_cell_voltage = true;
+        } else {
+            critical_cell_voltage = false;
+        }
+
+        // low cell voltage or highest to lowest diff too great
```
Comment (WickedShell) [inline libraries/AP_BattMonitor/AP_BattMonitor_Backend.cpp]: Cell spread should not overload the low voltage one, as it hides information from the caller/makes the overall display of failures more problematic.

File: libraries/AP_BattMonitor/AP_BattMonitor_SMBus_Solo.cpp (hunk 3 above)
Comment (Pedals2Paddles) [reply libraries/AP_BattMonitor/AP_BattMonitor_SMBus_Solo.cpp]: I would rather it out in the failsafe check too, but I couldn't see where I'd get a cell count to use in the for loop.  What do you suggest for that?

File: libraries/AP_BattMonitor/AP_BattMonitor.h (hunk 1 above)
Comment (Pedals2Paddles) [reply libraries/AP_BattMonitor/AP_BattMonitor.h]: Done

File: libraries/AP_BattMonitor/AP_BattMonitor_SMBus_Solo.cpp (hunk 3 above)
Comment (Pedals2Paddles) [reply libraries/AP_BattMonitor/AP_BattMonitor_SMBus_Solo.cpp]: Reverted. Will think of something else.

File: libraries/AP_BattMonitor/AP_BattMonitor.cpp (hunk 2 above)
Comment (Pedals2Paddles) [reply libraries/AP_BattMonitor/AP_BattMonitor.cpp]: Reverted. Will think of something else.

File: libraries/AP_BattMonitor/AP_BattMonitor_SMBus_Solo.cpp (hunk 3 above)
Comment (Pedals2Paddles) [reply libraries/AP_BattMonitor/AP_BattMonitor_SMBus_Solo.cpp]: Done. Moved out to the state.

File: libraries/AP_BattMonitor/AP_BattMonitor_Backend.cpp (hunk 4 above)
Comment (Pedals2Paddles) [reply libraries/AP_BattMonitor/AP_BattMonitor_Backend.cpp]: Still can't think of a better way that isn't a mess.  It won't override anything else. But I agree there must be a better way. And this kinda also goes back to the status text not being able to convey the full message.

**Comment by Pedals2Paddles** [comment]
@WickedShell

**Comment by Pedals2Paddles** [comment]
@WickedShell

File: libraries/AP_BattMonitor/AP_BattMonitor_Backend.cpp [hunk 5]
```diff
@@ -17,6 +17,7 @@
 #include <AP_HAL/AP_HAL.h>
 #include "AP_BattMonitor.h"
 #include "AP_BattMonitor_Backend.h"
+#include <GCS_MAVLink/GCS.h>
```
Comment (WickedShell) [inline libraries/AP_BattMonitor/AP_BattMonitor_Backend.cpp]: Why was this needed? Nothing jumps out at me as a missing feature.

File: libraries/AP_BattMonitor/AP_BattMonitor_Backend.cpp [hunk 6]
```diff
@@ -214,4 +217,47 @@ void AP_BattMonitor_Backend::check_failsafe_types(bool &low_voltage, bool &low_c
...
+    if (has_cell_voltages()) {
+
+        uint16_t cell_diff_mv = 0.0f;
+        uint16_t cell_lowest_mv = _state.cell_voltages.cells[0];  // initialize with the first cell's voltage
+        uint16_t cell_highest_mv = _state.cell_voltages.cells[0]; // initialize with the first cell's voltage
+
+        for (uint8_t i = 0; i < ARRAY_SIZE(_state.cell_voltages.cells); i++) {
+            
+            // bail out if cell voltage array element is UINT16_MAX since that means we've reached elements
+            if (_state.cell_voltages.cells[i] == UINT16_MAX) {
```
Comment (WickedShell) [inline libraries/AP_BattMonitor/AP_BattMonitor_Backend.cpp]: This isn't actually true. You need to loop through the whole array. On Maxell back ends we can lose the middle voltage and fill it in as `UINT16_MAX` if that happens. (The Maxell behavior is admittedly a bit questionable, but that is the current state of implementation.)

File: libraries/AP_BattMonitor/AP_BattMonitor_Backend.cpp [hunk 7]
```diff
@@ -214,4 +217,47 @@ void AP_BattMonitor_Backend::check_failsafe_types(bool &low_voltage, bool &low_c
     } else {
         low_capacity = false;
     }
+
+    // check battery cell voltages
+    if (has_cell_voltages()) {
+
+        uint16_t cell_diff_mv = 0.0f;
+        uint16_t cell_lowest_mv = _state.cell_voltages.cells[0];  // initialize with the first cell's voltage
+        uint16_t cell_highest_mv = _state.cell_voltages.cells[0]; // initialize with the first cell's voltage
```
Comment (WickedShell) [inline libraries/AP_BattMonitor/AP_BattMonitor_Backend.cpp]: This could be UINT16_MAX in which case you will always fail (in the case of the Maxell driver suffering from noise on the bus). Just start this out with 0, since you already loop through the first one anyways.

File: libraries/AP_BattMonitor/AP_BattMonitor_Backend.cpp [hunk 8]
```diff
@@ -214,4 +217,47 @@ void AP_BattMonitor_Backend::check_failsafe_types(bool &low_voltage, bool &low_c
...
+        // critical cell voltage
+        if (is_positive(_params._crt_cell_voltage) && ((float)(cell_lowest_mv * 1e-3) < _params._crt_cell_voltage)) {
+            critical_cell_voltage = true;
+        } else {
+            critical_cell_voltage = false;
+        }
+
+        // low cell voltage or cell highest to lowest diff
+        if ((is_positive(_params._low_cell_voltage) && ((float)(cell_lowest_mv * 1e-3) < _params._low_cell_voltage)) || 
+           (is_positive(_params._cell_voltage_diff) && ((float)(cell_diff_mv * 1e-3) > _params._cell_voltage_diff))) {
```
Comment (WickedShell) [inline libraries/AP_BattMonitor/AP_BattMonitor_Backend.cpp]: Add an extra boolean argument for the critical version. (Which allows you to have a proper error string for that).

Comment (WickedShell) [reply libraries/AP_BattMonitor/AP_BattMonitor_Backend.cpp]: Long term (but out of scope/not needed for this PR), I should really just make that a struct that we fill in, the list has gotten unwieldy long)

File: libraries/AP_BattMonitor/AP_BattMonitor_Backend.cpp (hunk 5 above)
Comment (Pedals2Paddles) [reply libraries/AP_BattMonitor/AP_BattMonitor_Backend.cpp]: Good question. Probably by accident. Deleting.

File: libraries/AP_BattMonitor/AP_BattMonitor_Backend.cpp (hunk 7 above)
Comment (Pedals2Paddles) [reply libraries/AP_BattMonitor/AP_BattMonitor_Backend.cpp]: OK

File: libraries/AP_BattMonitor/AP_BattMonitor_Backend.cpp (hunk 6 above)
Comment (Pedals2Paddles) [reply libraries/AP_BattMonitor/AP_BattMonitor_Backend.cpp]: lol ok

File: libraries/AP_BattMonitor/AP_BattMonitor_Backend.cpp (hunk 8 above)
Comment (Pedals2Paddles) [reply libraries/AP_BattMonitor/AP_BattMonitor_Backend.cpp]: Good point on the critical bool.  Done.

File: libraries/AP_BattMonitor/AP_BattMonitor_Backend.cpp [hunk 9]
```diff
@@ -214,4 +216,42 @@ void AP_BattMonitor_Backend::check_failsafe_types(bool &low_voltage, bool &low_c
...
     }
+
+    // check battery cell voltages
+    if (has_cell_voltages()) {
+
+        uint16_t cell_diff_mv = 0.0f;
+        uint16_t cell_lowest_mv = _state.cell_voltages.cells[0];  // initialize with the first cell's voltage
+        uint16_t cell_highest_mv = 0; // initialize with the first cell's voltage
+
+        for (uint8_t i = 0; i < ARRAY_SIZE(_state.cell_voltages.cells); i++) {
```
[the remark this answers is not in the document]
Comment (WickedShell) [reply libraries/AP_BattMonitor/AP_BattMonitor_Backend.cpp]: There isn't a configured cells count, you just find the number of cells by looping.

**Comment by Pedals2Paddles** [comment]
Dev call discussion:
* The cell difference check probably makes the low cell voltage check irrelevant.
* Cell difference check should either be critical only or have a low and critical level.
* Should we be using cell voltage rather than total pack voltage for the voltage the voltage failsafes?  IMO, yes.

**Comment by Pedals2Paddles** [comment]
I made some changes to this.  It will now only monitor the cell difference.  There are only two new parameters, one for a low and one for a critical difference.  I believe this accomplishes what every was looking for on the differential.  I think this is the big flight safety enhancement, and accomplishes what I was personally looking for too.

Regarding low cell voltage vs low pack total voltage, I'm happy save for another PR.  It's not necessary for this, and sounds like a separate problem to solve that is not longer related.

If everyone is happy with this, I will fixup the commits to make it merge ready.

**Comment by lvale** [comment]
Instead of working around the issue, I believe this should be used to redo the battery management.
Instead of inputting total battery values and defining the failsafe and critical values for the entire pack, the battery should be defined as a single cell, and the corresponding failsafe and critical value related to the single cell. What remains missing would be an extra config that would set the number of cells.

So internally the current system would work by setting #cells*voltages and the per cell alarm would also be valid....

**Comment by Pedals2Paddles** [comment]
As much as I would like this to be an ArduPilot function, this is simply way more work than it is worth now. It seems the Solo is the only vehicle widely using smart batteries _and_ desiring cell voltage monitoring _and_ wanting to make it happen.  So instead I solved this with about 5 minutes worth of python onboard the Solo's co-computer.

The hours upon hours spent redoing this over and over, more parameters, less parameters, reinvent this and that, all to do the same thing is simply not beneficial anymore.  Maybe when smart batteries become more prevalent, this will be a more worthwhile effort for someone.
